activation/nipost_test.go

```diff
@@ -807,6 +807,298 @@ func TestNIPSTBuilder_PoetUnstable(t *testing.T) {
+
+			existingRegistrations, err := nb.submitPoetChallenges(
+				context.Background(),
+				sig,
+				time.Now().Add(10*time.Second),
+				time.Now().Add(-5*time.Second), // poet round started
+				challengeHash.Bytes())
+
+			require.NoError(t, err)
+			require.Len(t, existingRegistrations, 1)
+			require.Equal(t, poetProverAddr, existingRegistrations[0].Address)
+		})
+
+	t.Run("1 poet removed AFTER round started -> too late to register to added poet",
+		func(t *testing.T) {
+			db := localsql.InMemory()
+			ctrl := gomock.NewController(t)
+
+			poetProver := NewMockPoetService(ctrl)
+			poetProver.EXPECT().Address().Return(poetProverAddr).AnyTimes()
+
+			addedPoetProver := NewMockPoetService(ctrl)
+			addedPoetProver.EXPECT().Address().Return(poetProverAddr2).AnyTimes()
+
+			// successfully registered to 2 poets
+			err = nipost.AddPoetRegistration(db, sig.NodeID(), nipost.PoETRegistration{
+				ChallengeHash: challengeHash,
+				Address:       poetProverAddr,
+				RoundID:       "1",
+				RoundEnd:      time.Now().Add(1 * time.Second),
+			})
+			require.NoError(t, err)
+
+			err = nipost.AddPoetRegistration(db, sig.NodeID(), nipost.PoETRegistration{
+				ChallengeHash: challengeHash,
+				Address:       poetProverAddr2,
+				RoundID:       "1",
+				RoundEnd:      time.Now().Add(1 * time.Second),
+			})
+
+			nb, err := NewNIPostBuilder(
+				db,
+				nil,
+				zaptest.NewLogger(t),
+				PoetConfig{},
+				nil,
+				nil,
+				WithPoetServices(poetProver),
+			)
+			require.NoError(t, err)
+
+			existingRegistrations, err := nb.submitPoetChallenges(
+				context.Background(),
+				sig,
+				time.Now().Add(10*time.Second),
+				time.Now().Add(-5*time.Second), // poet round started
+				challengeHash.Bytes())
+
+			require.NoError(t, err)
+			require.Len(t, existingRegistrations, 1)
+			require.Equal(t, poetProverAddr, existingRegistrations[0].Address)
+		})
+
+	t.Run("completely changed poet service AFTER round started -> fail, too late to register again",
+		func(t *testing.T) {
+			db := localsql.InMemory()
+			ctrl := gomock.NewController(t)
+
+			poetProver := NewMockPoetService(ctrl)
+			poetProver.EXPECT().Address().Return(poetProverAddr).AnyTimes()
+
+			// successfully registered to removed poet
+			err = nipost.AddPoetRegistration(db, sig.NodeID(), nipost.PoETRegistration{
+				ChallengeHash: challengeHash,
+				Address:       poetProverAddr2,
+				RoundID:       "1",
+				RoundEnd:      time.Now().Add(1 * time.Second),
+			})
+			require.NoError(t, err)
+
+			logger := zaptest.NewLogger(t)
+
+			nb, err := NewNIPostBuilder(
+				db,
+				nil,
+				logger,
+				PoetConfig{},
+				nil,
+				nil,
+				WithPoetServices(poetProver),
+			)
+			require.NoError(t, err)
+
+			_, err = nb.submitPoetChallenges(
+				context.Background(),
+				sig,
+				time.Now().Add(10*time.Second),
+				time.Now().Add(-5*time.Second), // poet round started
+				challengeHash.Bytes())
+			require.ErrorIs(t, err, ErrNoRegistrationForGivenPoetFound)
```

Comment:
Instead of `require.ErrorIs`, `require.ErrorAs` can be used here:

```suggestion
			poetErr := &PoetRegistrationMismatchError{}
			require.ErrorAs(t, err, &poetErr)
			require.ElementsMatch(t, []string{poetProverAddr2}, poetErr.registrations)
			require.ElementsMatch(t, []string{poetProverAddr}, poetErr.configuredPoets)
```